SVGPathElement.getTotalLength() should respect CSS 'd' property for display:none elements

Ahmad-S792 opened the pull request:
#### fb740fd4160c7a929a26ef4d16302a441f43919c
<pre>
SVGPathElement.getTotalLength() should respect CSS &apos;d&apos; property for display:none elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=304701">https://bugs.webkit.org/show_bug.cgi?id=304701</a>
<a href="https://rdar.apple.com/167195297">rdar://167195297</a>

Reviewed by NOBODY (OOPS!).

Per the SVG specification [1], getTotalLength() should compute the path length
based on the effective path data, which includes CSS &apos;d&apos; property values.
Currently, display:none elements fail to read the CSS &apos;d&apos; property because
pathByteStream() only checks renderer-&gt;style(), which is null for non-rendered
elements.

Fix by checking computedStyle() first before falling back to the renderer,
allowing display:none elements to access their CSS &apos;d&apos; property values.

[1] <a href="https://svgwg.org/svg2-draft/paths.html#InterfaceSVGGeometryElement">https://svgwg.org/svg2-draft/paths.html#InterfaceSVGGeometryElement</a>

* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-04-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGPathElement.getTotalLength-01-expected.txt: Ditto
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::pathByteStream const):
</pre>

annevk commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -238,6 +238,17 @@ RenderPtr<RenderElement> SVGPathElement::createElementRenderer(RenderStyle&& sty
 {
     if (document().settings().cssDPropertyEnabled()) {
+        // Ensure style is computed before accessing it.
+        protectedDocument()->updateStyleIfNeeded();
```
Why would we do this even when there is a renderer?

Ahmad-S792 replied:
Without this - when locally I ran `svg` test suite (both internal + wpt) that we were crashing few test cases, kind of like getting stale styles or from detached element. So I added it to ensure we update style before accessing them.

annevk replied:
But do we also need to update it when we have a renderer? Or only in the display:none case?

Ahmad-S792 replied:
Only in `display:none` case since we had the code with renderer without this update style without issue. Only with display:none case - where we don't have renderer and try to use `computedStyle`, it needs to be ensure that it is up-to-date, so we don't crash or get stale style.

annevk replied:
But now you are also computing it in the other branch. Do you see the issue?

Ahmad-S792 replied:
I don't see any issue, why this should be an issue for renderer case?

annevk replied:
Because we’re doing work that’s not needed?

nullhook replied:
@Ahmad-S792 you should only reflow when there's no renderer

nit: also, do the fast path first


```c++
if (CheckedPtr renderer = this->renderer()) {
  if (auto& pathFunction = renderer->style().d().tryPath())
     return pathFunction->parameters.data.byteStream;
  return SVGPathByteStream::empty();
}

protectedDocument()->updateStyleIfNeeded();
if (auto* style = computedStyle()) {
   if (auto& pathFunction = style->d().tryPath())
      return pathFunction->parameters.data.byteStream;
}
```

nullhook replied:
just curious, calling `computedStyle()` should typically reflow implicitly. why do need an explicit `updateStyleIfNeeded` here?

> we were crashing few test cases, kind of like getting stale styles or from detached element.

did you get crash/test failures without `updateStyleIfNeeded()` ?

Ahmad-S792 replied:
Yes - I got crashes without `updateStyleIfNeeded()` - hence I added it.

annevk replied:
Let me rephrase: check the renderer first. Add a fallback for computed style.

Ahmad-S792 committed:
SVGPathElement.getTotalLength() should respect CSS 'd' property for display:none elements
https://bugs.webkit.org/show_bug.cgi?id=304701
rdar://167195297

Reviewed by NOBODY (OOPS!).

Per the SVG specification [1], getTotalLength() should compute the path length
based on the effective path data, which includes CSS 'd' property values.
Currently, display:none elements fail to read the CSS 'd' property because
pathByteStream() only checks renderer->style(), which is null for non-rendered
elements.

Fix by checking computedStyle() first before falling back to the renderer,
allowing display:none elements to access their CSS 'd' property values.

[1] https://svgwg.org/svg2-draft/paths.html#InterfaceSVGGeometryElement

* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::pathByteStream const):
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-04-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGPathElement.getTotalLength-01-expected.txt: Ditto
* LayoutTests/platform/mac/TestExpectations:

nikolaszimmermann commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -238,6 +238,14 @@ RenderPtr<RenderElement> SVGPathElement::createElementRenderer(RenderStyle&& sty
 const SVGPathByteStream& SVGPathElement::pathByteStream() const
 {
     if (document().settings().cssDPropertyEnabled()) {
+        // Try computed style first (works for display:none elements).
```
In what situations are we still asking the renderer for it's style now? Shouldn't the computed style always be available? I'm confused, and maybe misunderstood something...

Ahmad-S792 replied:
@nikolaszimmermann - My understanding, when we don't have renderer display: none plus some display:cotents cases, we have to get computed style to give information - this is what the test is expecting.
